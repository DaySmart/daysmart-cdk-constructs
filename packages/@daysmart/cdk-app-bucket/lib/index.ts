import cdk = require('@aws-cdk/core');
import s3 = require('@aws-cdk/aws-s3');

export class AppBucket extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: any) {
        super(scope, id);

        new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.stage}-${props.appName.toLowerCase()}-website-bucket`,
            publicReadAccess: true,
            versioned: true,
        });
    }
}