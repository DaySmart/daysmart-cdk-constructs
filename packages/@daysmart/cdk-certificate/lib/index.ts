import * as cdk from '@aws-cdk/core';
import acm = require('@aws-cdk/aws-certificatemanager');
import route53 = require('@aws-cdk/aws-route53');

export interface CdkCertificateProps {
    domainName: string;
    stage: string;
    project: string;
}

export class CdkCertificate extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkCertificateProps) {
    super(scope, id);

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: props.domainName
    });

    new acm.Certificate(this, 'Certificate', {
        domainName: props.domainName,
        subjectAlternativeNames: [
            `${props.stage}.${props.project}.${props.domainName}`,
            `*.${props.stage}.${props.project}.${props.domainName}`
        ],
        validation: acm.CertificateValidation.fromDns(hostedZone)
    });
  }
}
