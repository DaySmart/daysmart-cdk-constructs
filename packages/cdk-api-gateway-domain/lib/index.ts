import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';

export interface CdkApiGatewayDomainProps {
  companyDomainName: string;
  companyHostedZoneId: string;
  project: string;
  baseEnv: string;
  dynamicEnv?: string;
  certificateArn: string;
  restApiId: string;
  basePath: string;
}

export class CdkApiGatewayDomain extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkApiGatewayDomainProps) {
    super(scope, id);

    const api = apigw.RestApi.fromRestApiId(this, "Rest API", `${props.restApiId}`);

    let customDomain: apigw.DomainName;

    if (props.dynamicEnv) {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: `${props.dynamicEnv}-api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    } else {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: (props.baseEnv == "prod") ? `api.${props.project}.${props.companyDomainName}` : `api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    }

    const cloudformationBasePathMapping = new apigw.CfnBasePathMapping(this, "CloudformationBasePathMapping", {
      basePath: `${props.basePath}`,
      domainName: customDomain.domainName,
      restApiId: `${props.restApiId}`,
      stage: `${props.baseEnv}` 
    });

    const cloudformationRoute53Change = new route53.CfnRecordSet(this, "CloudformationAliasRecordRoute53", {
      aliasTarget: {
        dnsName: customDomain.domainNameAliasDomainName,
        evaluateTargetHealth: false,
        hostedZoneId: "Z2FDTNDATAQYW2"
      },
      comment: `CloudFront distribution for ${customDomain.domainName}`,
      hostedZoneId: props.companyHostedZoneId,
      name: customDomain.domainName,
      type: "A"
    });
  }
}
