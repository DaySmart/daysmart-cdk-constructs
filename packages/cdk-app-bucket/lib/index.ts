import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudfront as cloudfront } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

export interface AppBucketProps {
    stage: string;
    appName: string;
    dynamicEnvName: string;
    projectName: string;
    sharedServicesAccountId?: string;
}
export class AppBucket extends Construct {
    constructor(scope: Construct, id: string, props: AppBucketProps) {
        super(scope, id);

        const bucket = new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.projectName}`,
            versioned: true,
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL
        });

        const originAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'OriginAcessIdentity', {
            comment: `OriginAccessIdentity for ${bucket.bucketName}.`
        });

        bucket.grantRead(originAccessIdentity)

        if(props.sharedServicesAccountId) {
            bucket.grantReadWrite(new iam.AccountPrincipal(props.sharedServicesAccountId));
        }

        let output = new cdk.CfnOutput(this, "AppBucket", {
            value: bucket.bucketName
        });

        let output2 = new cdk.CfnOutput(this, "OriginAccessIdentity", {
            value: originAccessIdentity.originAccessIdentityName
        });

        output.overrideLogicalId("AppBucket");
        output2.overrideLogicalId("OriginAccessIdentity");
    }
}