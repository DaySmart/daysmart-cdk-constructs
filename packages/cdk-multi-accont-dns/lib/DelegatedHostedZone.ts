import * as cdk from '@aws-cdk/core';
import * as route53 from '@aws-cdk/aws-route53';

/**
 * Properties for a hosted zone to be delegated
 */
export interface DelegatedHostedZoneProps {
    /**
     * Name of the hosted zone. i.e. example.domain.com
     */
    zoneName: string;
}

/**
 * A Route 53 hosted zone to be delegated from another account
 */
export class DelegatedHostedZone extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: DelegatedHostedZoneProps) {
        super(scope, id);

        const hostedZone = new route53.HostedZone(this, 'HostedZone', {
            zoneName: props.zoneName
        });

        const zoneId = new cdk.CfnOutput(this, 'HostedZoneId', {
            value: hostedZone.hostedZoneId
        });
        zoneId.overrideLogicalId('HostedZoneId');

        if(hostedZone.hostedZoneNameServers) {
            const nameServersList = hostedZone.hostedZoneNameServers as string[];
            const nameServers = new cdk.CfnOutput(this, 'NameServers', {
                value: cdk.Fn.join(',', nameServersList)
            });
            nameServers.overrideLogicalId('NameServers');
        }
    }

}