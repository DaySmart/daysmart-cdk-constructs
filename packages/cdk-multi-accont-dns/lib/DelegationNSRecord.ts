import * as cdk from '@aws-cdk/core';
import * as route53 from '@aws-cdk/aws-route53';

/**
 * Properties for an NS record to delegate a hosted zone
 */
export interface DelegatedNSRecordProps {
    /**
     * The hosted zone that requires the NS record. i.e domain.com
     */
    hostedZoneDomain: string;

    /**
     * The hosted zone to delegate. i.e. example.domain.com
     */
    delegatedDomainName: string;

    /**
     * A comma seperated list of name servers for the delegated hosted zone.
     * This can be derived from the NameServers output on the DelegatedHostedZone construct
     */
    nameServers: string;
}

/**
 * A Route 53 record set to delegate a hosted zone to another account
 */
export class DelegatedNSRecord extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: DelegatedNSRecordProps) {
        super(scope, id);

        const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
            domainName: props.hostedZoneDomain
        });

        new route53.NsRecord(this, 'NSRecordSet', {
            zone: hostedZone,
            recordName: props.delegatedDomainName,
            values: props.nameServers.split(',')
        });
    }
}