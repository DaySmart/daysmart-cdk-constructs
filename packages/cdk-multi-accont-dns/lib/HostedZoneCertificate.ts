import * as cdk from 'aws-cdk-lib'; 
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as route53 from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

/**
 * Properties for a ACM certificate for a hosted zone
 */
export interface HostedZoneCertificateProps {
    /**
     * The domain name of the hosted zone. i.e example.domain.com
     */
    domainName: string;
}

/**
 * An ACM certificate for all domains in a hosted zone
 * For example the domain example.domain.com would include:
 *  - example.domain.com
 *  - *.example.domain.com
 */
export class HostedZoneCertificate extends Construct {
    constructor(scope: Construct, id: string, props: HostedZoneCertificateProps) {
        super(scope, id);

        const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
            domainName: props.domainName
        });

        const certificate = new acm.DnsValidatedCertificate(this, 'Certificate', {
            hostedZone: hostedZone,
            domainName: props.domainName,
            subjectAlternativeNames: [`*.${props.domainName}`]
        });

        const certificateArn = new cdk.CfnOutput(this, 'CertificateArn', {
            value: certificate.certificateArn
        });
        certificateArn.overrideLogicalId('CertificateArn');
    }
}