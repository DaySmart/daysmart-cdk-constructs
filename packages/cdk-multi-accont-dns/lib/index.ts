export { DelegatedHostedZone, DelegatedHostedZoneProps } from './DelegatedHostedZone';
export { DelegatedNSRecord, DelegatedNSRecordProps } from './DelegationNSRecord';
export { HostedZoneCertificate, HostedZoneCertificateProps } from './HostedZoneCertificate'; 