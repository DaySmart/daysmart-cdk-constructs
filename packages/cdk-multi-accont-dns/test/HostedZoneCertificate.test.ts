import * as cdk from '@aws-cdk/core';
import '@aws-cdk/assert/jest'
import { SynthUtils } from '@aws-cdk/assert';
import { HostedZoneCertificate } from '../lib';

test('Hosted Zone certificate template', () => {
    const stack = new cdk.Stack(undefined, 'stack', {
        env: {
            account: '123456',
            region: 'us-east-1'
        }
    });
    new HostedZoneCertificate(stack, 'HostedZoneCert', {
        domainName: 'example.domain.com'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});