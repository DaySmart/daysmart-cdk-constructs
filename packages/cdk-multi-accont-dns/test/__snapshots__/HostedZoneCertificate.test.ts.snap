// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hosted Zone certificate template 1`] = `
Object {
  "Outputs": Object {
    "CertificateArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "HostedZoneCertCertificateCertificateRequestorResource118AF24B",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4ArtifactHash5E28809B": Object {
      "Description": "Artifact hash for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE": Object {
      "Description": "S3 bucket for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC": Object {
      "Description": "S3 key for asset version \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HostedZoneCertCertificateCertificateRequestorFunction39575E7A": Object {
      "DependsOn": Array [
        "HostedZoneCertCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy2FA8415F",
        "HostedZoneCertCertificateCertificateRequestorFunctionServiceRole8DD610F6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HostedZoneCertCertificateCertificateRequestorFunctionServiceRole8DD610F6",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HostedZoneCertCertificateCertificateRequestorFunctionServiceRole8DD610F6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HostedZoneCertCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy2FA8415F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/DUMMY",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HostedZoneCertCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy2FA8415F",
        "Roles": Array [
          Object {
            "Ref": "HostedZoneCertCertificateCertificateRequestorFunctionServiceRole8DD610F6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HostedZoneCertCertificateCertificateRequestorResource118AF24B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "example.domain.com",
        "HostedZoneId": "DUMMY",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "HostedZoneCertCertificateCertificateRequestorFunction39575E7A",
            "Arn",
          ],
        },
        "SubjectAlternativeNames": Array [
          "*.example.domain.com",
        ],
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
